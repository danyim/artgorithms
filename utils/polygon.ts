export interface Bounds {
  xMin: number;
  xMax: number;
  yMin: number;
  yMax: number;
}

export interface Point {
  x: number;
  y: number;
}

/** Parameters for fill/stokeRect */
export interface RectParam {
  x: number;
  y: number;
  w: number;
  h: number;
}

/** Random real number in a given range */
export const randRange = (min: number, max: number) =>
  Math.random() * (max - min) + min;

/** Random integer in a given range */
export const randIntegerRange = (min: number, max: number) =>
  Math.floor(randRange(min, max));

/** Returns the center of gravity of a polygon, i.e. centoid */
export const centroid = (vertices: Point[]): Point => {
  const x = (a, v: Point) => a + v.x;
  const y = (a, v: Point) => a + v.y;
  return {
    x: vertices.reduce(x, 0) / vertices.length,
    y: vertices.reduce(y, 0) / vertices.length,
  };
};

export const vector = (a: Point, b: Point) => ({
  x: b.x - a.x,
  y: b.y - a.y,
});

/** Dot product of two points */
export const dot = (u: Point, v: Point) => u.x * v.x + u.y * v.y;

/** Euclidean distance between two points */
export const distance = (a: Point, b: Point) =>
  Math.sqrt((a.x - b.x) ** 2 + (a.y - b.y) ** 2);

export const closestDistToPoint = (point: Point, pointsToPick: Point[]) => {
  let closestDistance: number = undefined;
  let closeastDistanceIndex;
  pointsToPick.forEach((pickPoint, index) => {
    const ptDistance = distance(point, pickPoint);

    if (!closestDistance || ptDistance < closestDistance) {
      closestDistance = ptDistance;
      closeastDistanceIndex = index;
    }
  });

  return pointsToPick[closeastDistanceIndex];
};

/** Pushes a value into an array if the value is truthy */
export function pushIfTruthy<A>(arr: A[], value: A) {
  if (value) {
    arr.push(value);
  }
}

/** Given certain parameters, this function will return the x,y positions of a wrapped row */
export const createWrappedRow = ({
  numItems,
  numPerLine,
  width,
  height,
  padding,
  offsetX,
  offsetY,
}: {
  numItems: number;
  numPerLine: number;
  width: number;
  height: number;
  padding: number;
  offsetX: number;
  offsetY: number;
}): Point[] => {
  // TODO: First position does _not_ start at 0,0 or offsetX,offsetY

  const positions: Point[] = [];
  let x = Math.max(offsetX - (width + padding), 0);
  let y = Math.max(offsetY - (width + padding), 0);
  let itemsPlaced = 0;
  for (let row = 0; row <= Math.ceil(numItems / numPerLine); row++) {
    for (let col = 0; col < numPerLine; col++) {
      if (itemsPlaced >= numItems) {
        break;
      }
      x += width + padding;
      itemsPlaced++;
      positions.push({
        x,
        y,
      });
    }
    x = Math.max(offsetX - (width + padding), 0);
    y += height + padding;
  }

  return positions;
};

/** Downsizes a square by a factor */
export const downsizeSquare = (
  squareSize: number,
  factor: number,
  offsetX: number = 0,
  offsetY: number = 0
): Point => {
  return {
    x: offsetX + (squareSize / 2 - (squareSize * factor) / 2),
    y: offsetY + (squareSize / 2 - (squareSize * factor) / 2),
  };
};

/** Returns the parameters to draw a square centered and scaled from the center as the origin (as
 * opposed to the top left) */
export const centerSquare = (x: number, y: number, size: number): RectParam => {
  return {
    x: x - size / 2,
    y: y - size / 2,
    w: size,
    h: size,
  };
};
