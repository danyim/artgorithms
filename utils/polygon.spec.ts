import { createWrappedRow, getBoundsCenter } from "./polygon";

describe("polygon utils", () => {
  describe("createWrappedRow", () => {
    it("should create a wrapped row", () => {
      const input = {
        numItems: 20,
        numPerLine: 5,
        width: 100,
        height: 100,
        padding: 10,
        offsetX: 0,
        offsetY: 0,
      };
      expect(createWrappedRow(input)).toMatchSnapshot();
    });

    it("should create a wrapped row starting at (0,0)", () => {
      const input = {
        numItems: 2,
        numPerLine: 1,
        width: 100,
        height: 100,
        padding: 10,
        offsetX: 0,
        offsetY: 0,
      };
      const result = createWrappedRow(input);
      expect(result[0]).toEqual({ x: 0, y: 0 });
      expect(result[1]).toEqual({ x: 0, y: input.height + input.padding });
    });

    it("should create a wrapped row starting at (100,100)", () => {
      const input = {
        numItems: 2,
        numPerLine: 1,
        width: 100,
        height: 100,
        padding: 10,
        offsetX: 100,
        offsetY: 100,
      };
      const result = createWrappedRow(input);
      expect(result[0]).toEqual({ x: 100, y: 100 });
      expect(result[1]).toEqual({ x: 100, y: 100 + 100 + 10 });
    });

    it("should not return more points than numItems", () => {
      const input = {
        numItems: 20,
        numPerLine: 5,
        width: 100,
        height: 100,
        padding: 10,
        offsetX: 0,
        offsetY: 0,
      };
      expect(createWrappedRow(input).length).toEqual(input.numItems);
    });
  });

  describe("getBoundsCenter", () => {
    it("should return the center", () => {
      const input = { xMin: 0, xMax: 10, yMin: 0, yMax: 10 };
      const expected = { x: 5, y: 5 };
      const result = getBoundsCenter(input);
      expect(result).toEqual(expected);
    });

    it("should return the center with an offset", () => {
      const input = { xMin: 4, xMax: 50, yMin: 15, yMax: 25 };
      const expected = { x: 27, y: 20 };
      const result = getBoundsCenter(input);
      expect(result).toEqual(expected);
    });
  });
});
