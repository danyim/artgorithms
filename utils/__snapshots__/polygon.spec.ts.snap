// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polygon utils createWrappedRow should create a wrapped row 1`] = `
[
  {
    "x": 0,
    "y": 0,
  },
  {
    "x": 110,
    "y": 0,
  },
  {
    "x": 220,
    "y": 0,
  },
  {
    "x": 330,
    "y": 0,
  },
  {
    "x": 440,
    "y": 0,
  },
  {
    "x": 0,
    "y": 110,
  },
  {
    "x": 110,
    "y": 110,
  },
  {
    "x": 220,
    "y": 110,
  },
  {
    "x": 330,
    "y": 110,
  },
  {
    "x": 440,
    "y": 110,
  },
  {
    "x": 0,
    "y": 220,
  },
  {
    "x": 110,
    "y": 220,
  },
  {
    "x": 220,
    "y": 220,
  },
  {
    "x": 330,
    "y": 220,
  },
  {
    "x": 440,
    "y": 220,
  },
  {
    "x": 0,
    "y": 330,
  },
  {
    "x": 110,
    "y": 330,
  },
  {
    "x": 220,
    "y": 330,
  },
  {
    "x": 330,
    "y": 330,
  },
  {
    "x": 440,
    "y": 330,
  },
]
`;
